### ISSUE TITLE
Crash Assembly$$GetAllAssemblies

### ISSUE THREAD

**Issue by auag0** [ByNameModding/BNM-Android#162]
ByNameModding: 2.5.2
game unity version: 6000.0.53f1

replaced
```
#define UNITY_VER 232 // 2023.2.x+

#define BNM_dlopen xdl_open
#define BNM_dlsym xdl_sym
#define BNM_dlclose xdl_close
#define BNM_dladdr xdl_addr

// dummy to dobby
```

code
```
#include <jni.h>
#include <string.h>
#include <thread>
#include <unistd.h>
#include "xdl.h"
#include "BNM/Loading.hpp"
#include "logger.h"

bool isSoLoaded(const char *so_name) {
    FILE *fp = fopen("/proc/self/maps", "r");
    if (!fp) {
        return false;
    }
    char line[512];
    bool loaded = false;
    while (fgets(line, sizeof(line), fp)) {
        if (strstr(line, so_name)) {
            loaded = true;
            break;
        }
    }
    fclose(fp);
    return loaded;
}

JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *vm, [[maybe_unused]] void *reserved) {
    JNIEnv *env;
    vm->GetEnv((void **) &env, JNI_VERSION_1_6);

    std::thread([]() {
        while (!isSoLoaded("libil2cpp.so")) {
            usleep(1000);
        }

        void *handle = xdl_open("libil2cpp.so", XDL_DEFAULT);
        LOGD("handle: %p", handle);
        BNM::Loading::TryLoadByDlfcnHandle(handle);
    }).detach();

    return JNI_VERSION_1_6;
}
```

Internals.cpp
```
IL2CPP::Il2CppImage *Internal::TryGetImage(const std::string_view &_name) {
    BNM_LOG_DEBUG("TryGetImage: %s", _name.data()); // logged
    auto &assemblies = *Internal::il2cppMethods.Assembly$$GetAllAssemblies(); // crash here maybe
    BNM_LOG_DEBUG("assemblies: %zu", assemblies.size()); // not logged
    for (auto assembly : assemblies) {
        auto currentImage = Internal::il2cppMethods.il2cpp_assembly_get_image(assembly);
        if (!Internal::CompareImageName(currentImage, _name)) continue;
        return currentImage;
    }

    return {};
}
```

Logcat
```
handle: 0x799e17df30
[SetupBNM] il2cpp::vm::Class::Init in lib: 0x76537d3614.
[INIT_IL2CPP_API]: il2cpp_image_get_class (il2cpp_image_get_class) in lib 0xfffffffed4ae35dc
[INIT_IL2CPP_API]: il2cpp_get_corlib (il2cpp_get_corlib) in lib 0xfffffffed4ae11b0
[INIT_IL2CPP_API]: il2cpp_class_from_name (il2cpp_class_from_name) in lib 0xfffffffed4ae1398
[INIT_IL2CPP_API]: il2cpp_assembly_get_image (il2cpp_assembly_get_image) in lib 0xfffffffed4ae12bc
[INIT_IL2CPP_API]: il2cpp_method_get_param_name (il2cpp_method_get_param_name) in lib 0xfffffffed4ae2ab4
[INIT_IL2CPP_API]: il2cpp_class_from_il2cpp_type (il2cpp_class_from_il2cpp_type) in lib 0xfffffffed4ae1390
[INIT_IL2CPP_API]: il2cpp_array_class_get (il2cpp_array_class_get) in lib 0xfffffffed4ae11dc
[INIT_IL2CPP_API]: il2cpp_type_get_object (il2cpp_type_get_object) in lib 0xfffffffed4ae3298
[INIT_IL2CPP_API]: il2cpp_object_new (il2cpp_object_new) in lib 0xfffffffed4ae2e28
[INIT_IL2CPP_API]: il2cpp_value_box (il2cpp_value_box) in lib 0xfffffffed4ae2e54
[INIT_IL2CPP_API]: il2cpp_array_new (il2cpp_array_new) in lib 0xfffffffed4ae1288
[INIT_IL2CPP_API]: il2cpp_field_static_get_value (il2cpp_field_static_get_value) in lib 0xfffffffed4ae2148
[INIT_IL2CPP_API]: il2cpp_field_static_set_value (il2cpp_field_static_set_value) in lib 0xfffffffed4ae2150
[INIT_IL2CPP_API]: il2cpp_string_new (il2cpp_string_new) in lib 0xfffffffed4ae2fe4
[INIT_IL2CPP_API]: il2cpp_resolve_icall (il2cpp_resolve_icall) in lib 0xfffffffed4ae11c0
[INIT_IL2CPP_API]: il2cpp_runtime_invoke (il2cpp_runtime_invoke) in lib 0xfffffffed4ae2f70
[INIT_IL2CPP_API]: il2cpp_domain_get (il2cpp_domain_get) in lib 0xfffffffed4ae1c24
[INIT_IL2CPP_API]: il2cpp_thread_current (il2cpp_thread_current) in lib 0xfffffffed4ae3078
[INIT_IL2CPP_API]: il2cpp_thread_attach (il2cpp_thread_attach) in lib 0xfffffffed4ae3088
[INIT_IL2CPP_API]: il2cpp_thread_detach (il2cpp_thread_detach) in lib 0xfffffffed4ae308c
[INIT_IL2CPP_API]: il2cpp_gc_alloc_fixed (il2cpp_gc_alloc_fixed) in lib 0xfffffffed4ae24d8
[INIT_IL2CPP_API]: il2cpp_gc_free_fixed (il2cpp_gc_free_fixed) in lib 0xfffffffed4ae24e0
[SetupBNM] code has il2cpp_image_get_class. BNM will use it.
[SetupBNM] il2cpp::vm::Class::FromIl2CppType in lib: 0xfffffffed44b938c.
[SetupBNM] il2cpp::vm::Type::GetClassOrElementClass in lib: 0xfffffffed44b938c.
[SetupBNM] il2cpp::vm::Image::FromName in lib: 0xfffffffed44abbc0.
[SetupBNM] il2cpp::vm::Assembly::GetAllAssemblies in lib: 0x76537b1490.
TryGetImage: UnityEngine.CoreModule.dll
```

**Comment by auag0** [comment]
sorry, i think its a problem with game

**Comment by auag0** [comment]
frida il2cpp bridge worked fine but not with lsposed/zygisk using bnm, dobby and xdl. Tried jni and handle for bnm loading

I don't know, but I'm using android 16

**Comment by auag0** [comment]
It now works after rewriting Assembly$$GetAllAssemblies to il2cpp_domain_get_assemblies.

**Comment by doviettung96** [comment]
@auag0 hi. I have a similar symptom that frida-il2cpp-bridge works for all of my codes but here some of my functions work, some failed.. not related to return values nor the arguments. They just work and fail :(. 
Anyway, why do you shift from frida-il2cpp-bridge to this? Similar reason to me?

**Comment by auag0** [comment]
@doviettung96 Hello
I wanted to use ImGui for my menu.
Frida can implement a UI as well, but it's less flexible and more troublesome.
Also, it's convenient that I can inject without needing a PC or Termux.
In my case, I use Frida for behavior checks and tracing during analysis, and then I implement everything with bnm/dobby.

**Comment by auag0** [comment]
Frida can catch most exceptions
but C++ doesn't work so well.
But in my case, I use BNM's Try catch or Try Invoke and it works fine.

**Comment by doviettung96** [comment]
oh thank you. yeah.. I can resolve my problem by attaching/ detaching il2cpp thread. but still, I got a new function that in frida I need to place within this
    await Il2Cpp.mainThread!.schedule(() => {
But it failed here (with the attach/detach il2cpp). The game just crashes without any exception.. just like before. Anyway, the attach/detach help me to run a lot il2cpp functions.. just now I got this new case. I bet I have to fix it next
